File: Problem_Solvings/백준/백트래킹/1987/1987.py
def bfs(current_row,current_col):
    q=set()
    q.add((current_row,current_col,arrays[current_row][current_col]))
    global max_distance
    max_distance=1
    while q:
        current_row,current_col,path=q.pop()
        max_distance=max(len(path),max_distance)
        for dir in range(4):
            next_row,next_col=current_row+row_move[dir],current_col+col_move[dir]
            if 0<=next_row<n_rows and 0<=next_col<n_cols:
                if arrays[next_row][next_col] not in path:
                    q.add((next_row,next_col,path+arrays[next_row][next_col]))

n_rows, n_cols=map(int,input().split(' '))
arrays=[]
for _ in range(n_rows):
    arrays.append(input())

row_move=[-1,0,1,0]
col_move=[0,1,0,-1]

bfs(0,0)
print(max_distance)

File: Problem_Solvings/백준/백트래킹/1987/1987.c
#define _CRT_SECURE_NO_WARNINGS
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#define TRUE 1
#define FALSE 0

int max_distance = 1;
char** arrays;
char path[27];

int dir_x[4] = { 0,1,0,-1 };
int dir_y[4] = { -1,0,1,0 };
int n_rows, n_cols;

int promising(char next_path, int distance) {
	int i;
	for (i = 0; i < distance; i++) {
		if (path[i] == next_path)
			return FALSE;
	}
	return TRUE;
}

void dfs(int current_row, int current_col, int distance) {
	int i, next_row, next_col;
	char next_char;
	if (distance > max_distance)
		max_distance = distance;

	for (i = 0; i < 4; i++) {
		next_row = current_row + dir_y[i];
		next_col = current_col + dir_x[i];

		if (next_row >= 0 && next_row < n_rows && next_col >= 0 && next_col < n_cols) {
			next_char = arrays[next_row][next_col];
			if (promising(next_char, distance)) {
				path[distance] = next_char;
				dfs(next_row, next_col, distance + 1);
			}

		}
	}
	return;
}
int main(void) {
	int i, num;
	scanf("%d %d", &n_rows, &n_cols);
	arrays = (char**)malloc(sizeof(char*)*n_rows);
	for (i = 0; i < n_rows; i++) {
		arrays[i] = (char*)malloc(sizeof(char)*n_cols);
		scanf("%s", arrays[i]);
	}
	path[0] = arrays[0][0];
	dfs(0, 0, 1);
	printf("%d", max_distance);

	for (i = 0; i < n_rows; i++)
		free(arrays[i]);
	free(arrays);
}